### DISCUSSION TITLE
Running a list of payloads

### DISCUSSION THREAD

**Discussion by kai-iak** [usebruno/bruno#1918] [Q&A] [answered]
Hi All:

Really enjoying Bruno so far.

I have a group of sample payloads that are used to test various scenarios of a POST.
Is there a way to loop through this list to test each post or do I need to create a new request for each payload?

Thanks!
K

**Comment by yannick-Lapierre** [comment]
I have a similar question.
I would like to run the same API endpoint, with a different value of 1 parameter. Would it be possible to do that without having to create a 1 query for each?

**Comment by tonkado** [comment]
Same question.
I've tried different approaches, but with no luck currently.

Here is my test:
```javascript
test("all categories are not empty", async function () {
  const axios = require("axios");
  const responseBody = res.getBody();
  const jsonFeedURL = responseBody.jsonFeedURL;
  
  for (const item in responseBody.cards) {
    const card = responseBody.cards[item];
    console.log(card.name);
    const url = jsonFeedURL + card.name;
    console.log(url);
    const response = await axios.get(url);
    console.log(response.data[0].id);
    expect(response.data[0].id).to.be.not.empty;
  }
});
```
As far as I understand, there is some issue with calling **expect** in the loop.
It brakes the loop but doesn't trigger the test to return false.

I believe there should be a solution for this. Any ideas?

**Comment by mrtfn** [comment]
Any luck on this?

**Reply by tonkado** [reply]
After your reply I went to check what is possible and found a solution.
The problem that it seems that Bruno totally ignore await and close the process without waiting.
Solution: move this kind of checks to Script section and use Pre Request or Post Response.

In my case I'm using Post Response:
```javascript
  const axios = require("axios");
  const responseBody = res.getBody();
  const jsonFeedURL = responseBody.jsonFeedURL;
  
  for (const item in responseBody.cards) {
    const card = responseBody.cards[item];
    const url = jsonFeedURL + card.name;
    const response = await axios.get(url);
    expect(response.data, "card [" + card.name + "] is empty").to.be.an('array').that.is.not.empty;
  }
```

**Reply by mrtfn** [reply]
Thank you for the fast response, but I may use some more of your help on this since JS is not my dream language.
The thing I want to do is that assume I have :
```json
{
    "id": 1,
    "name": "Leanne Graham",
    "username": "Bret",
    "email": "Sincere@april.biz",
    "address": {
      "street": "Kulas Light",
      "suite": "Apt. 556",
      "city": "Gwenborough",
      "zipcode": "92998-3874",
      "geo": {
        "lat": "-37.3159",
        "lng": "81.1496"
      }
    },
    "phone": "1-770-736-8031 x56442",
    "website": "hildegard.org",
    "company": {
      "name": "Romaguera-Crona",
      "catchPhrase": "Multi-layered client-server neural-net",
      "bs": "harness real-time e-markets"
    }
  }
```
And I want to test my API with different scenarios to trigger validation errors, using approximately 100 alternatives of the above JSON for this purpose.
Is this the case? Should I or can I save these somewhere easily accessible in Bruno, or just paste them in the test section and loop through it ? Or im completely off the road for my purpose(maybe a separate issue or discussion, idk, guide me if you can, illl be grateful)?

**Reply by tonkado** [reply]
I'm not a great expert either 😁
If I needed to solve this task, I would make one of these:

1. Add all 100 cases as an array in a single JSON file and host it somewhere.
Then used axios to get this JSON with all cases and loop through them (as described above) in the Post Response section.

2. Add it as a module to store with Bruno test files (doc: https://docs.usebruno.com/testing/script/js-file).
And then just include this module with all test cases and run them in the loop.

You can ask ChatGPT or any other LLM for some ideas and find the solution that works best for you.
Good luck!

**Reply by mrtfn** [reply]
Thank you for the suggestion! That’s a good option. I'm considering implementing an endpoint suffix for each endpoint, so I could access the payloads via something like `api.local/register/test` and then use them on `api.local/register`. I’m not sure if it’s a great idea yet, but I’ll think about it later xd!

**Reply by helloanoop** [reply]
@mrtfn [Data Driven testing](https://docs.usebruno.com/testing/automate-test/data-driven-testing) might help with what you are describing.

**Comment by helloanoop** [comment, marked as answer]
To run through a list of payloads, we recommend storing them in JSON/CSV and iterating through them in the collection runner.
See [Data Driven testing](https://docs.usebruno.com/testing/automate-test/data-driven-testing) for more details

**Reply by mrtfn** [reply]
Thanks for clarifying, I haven't noticed that feature in docs, btw since it's behind a paywall I would go with another solution like the one provided earlier by @tonkado to see if it works for my hobby project
